scipy/stats/_survival.py

```diff
@@ -1,38 +1,213 @@
+        self._sf = points if kind == 'sf' else 1 - points
+        self._kind = kind
+        f0 = 1 if kind == 'sf' else 0  # leftmost function value
+        f1 = 1 - f0
+        # fill_value can't handle edge cases at infinity
+        x = np.insert(x, [0, len(x)], [-np.inf, np.inf])
+        y = np.insert(points, [0, len(points)], [f0, f1])
+        # `or` conditions handle the case of empty x, points
+        self._f = interpolate.interp1d(x, y, kind='previous',
+                                       assume_sorted=True)
+
+    def confidence_interval(self, confidence_level=0.95, method='Greenwood'):
+        """Compute a confidence interval around the CDF/SF point estimate
+
+        Parameters
+        ----------
+        confidence_level : float, default: 0.95
+            Confidence level for the computed confidence interval
+
+        method : str, optional
+            Method used to compute the confidence interval. Options are
+            "Greenwood" and "Exponential Greenwood".
+
+        Returns
+        -------
+        ci : ``ConfidenceInterval``
+            An object with attributes ``low`` and ``high``: arrays of the
+            lower and upper bounds of the confidence interval at unique values
+            of the sample.
+
+        Notes
+        -----
+        Confidence intervals are computed according to the Greenwood formula or
+        the more recent "Exponential Greenwood" formula as described in [1]_.
+
+        References
+        ----------
+        .. [1] Sawyer, Stanley. "The Greenwood and Exponential Greenwood
+               Confidence Intervals in Survival Analysis."
+               https://www.math.wustl.edu/~sawyer/handouts/greenwood.pdf
+
+        """
+        methods = {'greenwood': self._greenwood_ci,
+                   'exponential greenwood': self._exp_greenwood_ci}
+
+        message = f"`method` must be one of {set(methods)}."
+        if method.lower() not in methods:
+            raise ValueError(message)
+
+        message = "`confidence_level` must be a scalar between 0 and 1."
+        confidence_level = np.asarray(confidence_level)[()]
+        if confidence_level.shape or not (0 <= confidence_level <= 1):
+            raise ValueError(message)
+
+        method_fun = methods[method.lower()]
+        low, high = method_fun(confidence_level)
+        return ConfidenceInterval(np.clip(low, 0, 1), np.clip(high, 0, 1))
+
+    def _greenwood_ci(self, confidence_level):
+        sf, d, n = self._sf, self._d, self._n
+        # When n == d, Greenwood's formula divides by zero.
+        # When s != 0, this can be ignored: var == inf, and CI is [0, 1]
+        # When s == 0, this results in NaNs. Produce an informative warning.
+        with np.errstate(divide='ignore', invalid='ignore'):
+            var = sf ** 2 * np.cumsum(d / (n * (n - d)))
+
+        message = ("The variance estimate is undefined at some observations. "
+                   "This is feature of the mathematical formula.")
```

Comment:
Nitpick to move that in the condition itself. Optional.

For the msg, I don't understand the formulation: `This is feature of`. Missing `a`?

Reply:
Yes, "a feature".
```suggestion
        message = ("The variance estimate is undefined at some observations. "
                   "This is a feature of the mathematical formula.")
```
Sometimes I intentionally take the message out of the branch. It documents what the branch is about.

Reply:
Corrected.